File: a6_p2.py
import numpy as np
import matplotlib.pyplot as plt
import sympy as sp


g = 9.81
h, u = sp.symbols('h u')

m = sp.Matrix([[0,1],[g*h-u**2, 2*u]])
eigenvalues = m.eigenvals()

e1, e2 = list(eigenvalues)


n = 500
a = 0
b = 5
dx = (b-a) / n
x = np.linspace(a, b, n+1)

T = 1
time = [0]


def spectral_rad(t):
    max_val = -1
    for i in range(n+1):
        temp_max = max(abs(u[i,t]+np.sqrt(g*h[i,t])), abs(u[i,t]-np.sqrt(g*h[i,t])))
        max_val = max(temp_max, max_val)
    return max_val  

# u(x, t)
u = np.zeros((n+1, 1000))
# h(x, t)
h = np.zeros((n+1, 1000))

for i in range(n+1):
    if((a+i*dx)<2.5) : h[i, 0] = 2
    else : h[i, 0] = 1
    u[i ,0] = 0


time_val = 0
for t in range(1001):
    dt = (0.9*dx)/(spectral_rad(t))
    time_val = time_val + dt
    if time_val > T:
        break
    time.append(time[-1]+dt)
    for i in range(n+1):
        l = i-1
        r = i+1
        if l == -1:
            l = 0
        if r == n+1:
            r = n
        
        h[i, t+1] = (h[r, t] + h[l, t])/2  - dt*(h[r, t]*u[r,t] - h[l,t]*u[l,t])/(2*dx) 
        u[i, t+1] = (h[r, t]*u[r,t] + h[l,t]*u[l,t])/(2*h[i, t+1]) - dt*(h[r,t]*u[r,t]**2+0.5*g*h[r,t]**2 - (h[l,t]*u[l,t]**2+0.5*g*h[l,t]**2))/(2*dx*h[i, t+1]) 
    
time_val = time_val - dt
print(time_val)
m = len(time)-1

plt.plot(x, u[:, m])
plt.show()
plt.plot(x, h[:, m])
plt.show()
'''
for t in range(m+1):
    plt.plot(x, u[:, t])
plt.show()

for t in range(m+1):
    plt.plot(x, h[:, t])
plt.show()
'''

File: a6_p2.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

const double g = 9.81;
const int n = 500;
const double a = 0;
const double b = 5;
const double dx = (b - a) / n;
const double T = 1.0;

std::vector<double> x(n + 1);

double spectral_rad(int t, const std::vector<std::vector<double>>& u, const std::vector<std::vector<double>>& h) {
    double max_val = -1;
    for (int i = 0; i <= n; ++i) {
        double lb1 = std::abs(u[i][t]+std::sqrt(g * h[i][t]));
        double lb2 = std::abs(u[i][t]-std::sqrt(g * h[i][t]));
        double temp_max = std::max(lb1, lb1);
        max_val = std::max(temp_max, max_val);
    }
    //std::cout<<max_val<<" ";
    return max_val;
}

int main() {
    std::vector<std::vector<double>> u(n + 1, std::vector<double>(1001, 0));
    std::vector<std::vector<double>> h(n + 1, std::vector<double>(1001, 0));

    // Set initial conditions for h(x, 0) and u(x, 0)
    for (int i = 0; i <= n; ++i) {
        x[i] = a + i * dx;
        if(x[i]<2.5) h[i][0] = 2;
        else h[i][0] = 1;
    }

    std::vector<double> time = {0};
    double time_val = 0;

    for (int t = 0; t <= 1000; ++t) {
        double dt = (0.9 * dx) / spectral_rad(t, u, h);
        //std::cout<<dt<<" ";
        time_val += dt;
        
        if (time_val > T) {
            break;
        }
        
        time.push_back(time.back() + dt);

        for (int i = 0; i <= n; ++i) {
            int l = i - 1;
            int r = i + 1;
            if (l == -1) l = 0;
            if (r == n + 1) r = n;

            h[i][t + 1] = (h[r][t] + h[l][t]) / 2 - dt * (h[r][t] * u[r][t] - h[l][t] * u[l][t]) / (2 * dx);
            u[i][t + 1] = (h[r][t] * u[r][t] + h[l][t] * u[l][t]) / (2 * h[i][t + 1]) - 
                dt * (h[r][t] * std::pow(u[r][t], 2) + 0.5 * g * std::pow(h[r][t], 2) -
                (h[l][t] * std::pow(u[l][t], 2) + 0.5 * g * std::pow(h[l][t], 2))) / (2 * dx * h[i][t + 1]);
        }
    }
    
    for(int t=0;t<time.size();t++) {
        for(int j=0;j<=n;j++) {
            std::cout<<"At time t="<<time[t]<<", x="<<x[j]<<" u="<<u[j][t]<<" and h="<<h[j][t]<<'\n';
        }
    }

    return 0;
}